### ISSUE TITLE
False positive with F401 (imported but unused)

### ISSUE THREAD

**Issue by sbdchd** [charliermarsh/ruff#1619]
minimal example:
```python
from typing import List

x = [] # type: List[str]
x.append("foo")
print(x)
```

This isn't a big deal since `# type: List[str]` can be replaced with `# type: list[str]` and mypy warns about `List` being undefined with the auto fixer removes the import

**Comment by charliermarsh** [comment]
Is this in reference to the type comment? Like, that `# type: List[str]` should be considered a valid reference to `List`? Just verifying.

**Comment by sbdchd** [comment]
Yeah exactly!

**Comment by charliermarsh** [comment]
Yeah we don't really support type comments right now, since they were made obsolete in Python 3.6. I know some projects still use them for compatibility... I'll think on it.

**Comment by spaceone** [comment]
Maybe the issue title can be change to `False positive with F401 (imported but unused) with mypy type comments` ? can faster be found when searching for this issue.

Affects me as well.

**Comment by uranusjr** [comment]
I’d say the comment looking nicer than a type annotation is an editor/highlighter problem, but the general point of this being useful is correct.

**Comment by johnthagen** [comment]
Similar to https://github.com/charliermarsh/ruff/issues/1619#issuecomment-1406845028, I still use type comments in some very rare cases where things like `for` loops don't have syntax to express a "real" Python 3 type hint and the type checkers can't infer the type due to limitations in the underlying libraries being used.

```py
from models import Wheel

for wheel in car.wheels.all():  # type: Wheel
    ...
```

I realize this is non-standard, but PyCharm will interpret this and provide intellisense/auto-complete for the loop variable, which is practically useful in these niche cases.

**Comment by johnthagen** [comment]
@JonathanPlasse Good point bringing that up. I've generally been a little resistant to this particular fix because it complicates the actual code just to add the hint:

- It creates a new unbound local (if the code gets refactored later it could open up the possibility that `s3: S3ServiceResource` gets "separated" from the loop accidentally)
- Python leaks iterators from the scope already, but the new `s3: S3ServiceResource` could be misunderstood as _implying intent_ to leak the iterator
- There is a DRY issue where if the loop variable `s3` is renamed but the outer typed variable isn't

But you're right that this is the officially prescribed way of dealing with it, but hopefully the above points help explain why people may be hesitant to do it just for a hint.

**Comment by V3RGANz** [comment]
Also, it's worth mentioning that if you unpack values in this manner:
```python
name: str
p: float
r: float
f1: float
for n, p, r, f1 in zip(names, precision, recall, f1_micro):
    ...
```
it seems unnecessarily complicated for such a simple functionality compared to a simpler approach like:
```python
for n, p, r, f1 in zip(names, precision, recall, f1_micro):  # type: str, float, float, float
    ...
```

**Comment by jaraco** [comment]
I stumbled onto this issue in the configparser project. I worked around it by converting the type comments to native type hints (jaraco/configparser@4990ba132450e218d9695ba74d4a139913459b17).

**Comment by charliermarsh** [comment]
To be clear, "obsolete" is different from "unused". I'm not claiming that type comments aren't used in existing projects or that type checkers do not support them. It would be nice to support them! But it's a matter of prioritization, and my understanding is that [PEP 526](https://peps.python.org/pep-0526/) had the explicit goal of introducing "a more readable syntax to replace type comments."

Perhaps another way of framing my lack of urgency to prioritize this: is there any reason to use type comments when writing Python code _today_? (This is not a rhetorical question, I am genuinely asking.)

**Comment by jaymegordo** [comment]
@charliermarsh personally I use them exclusively, (see my previous comment https://github.com/astral-sh/ruff/issues/1619#issuecomment-1518178907), purely because I think they make code much more readable. But thats just my 2c 🤷🏼‍♂️

**Comment by johnthagen** [comment]
@charliermarsh 

> is there any reason to use type comments when writing Python code today?

As explained here, for the rare cases of needing to type hint loop variables, I find type comments avoid several drawbacks:

- https://github.com/astral-sh/ruff/issues/1619#issuecomment-1557012758

**Comment by charliermarsh** [comment]
(Thank you, these comments are all welcome and helpful.)

**Comment by johnthagen** [comment]
pyflakes used to support this (used it for years). Perhaps there was a regression.

- https://github.com/PyCQA/pyflakes/pull/400

**Comment by charliermarsh** [comment]
It looks like it was removed about a year ago: https://github.com/PyCQA/pyflakes/pull/684

**Comment by ndevenish** [comment]
We have some scripts that we explicitly maintain compatibility with python 2 (generally bootstrapping-related scripts that people can run on python2 and take care of bootstrapping a full environment).

This is _very mildly_ annoying but easily fixed with a `noqa: F401` on the import, and is a narrow enough use case that this causes no pain.
